frame/contracts/src/lib.rs

```diff
@@ -955,9 +957,70 @@ pub mod pallet {
 		StorageValue<_, BoundedVec<DeletedContract, T::DeletionQueueDepth>, ValueQuery>;
 }
 
+/// The type of origins supported by the contracts pallet.
+#[derive(Clone)]
+pub enum ContractOrigin<T: Config> {
+	/// The origin is a regular signed account.
+	Signed(T::AccountId),
+	/// The origin is root.
+	Root,
+}
+
+impl<T: Config> ContractOrigin<T> {
+	/// Creates a new Signed ContractOrigin from an AccountId.
+	pub fn from_account_id(account_id: T::AccountId) -> Self {
+		ContractOrigin::Signed(account_id)
+	}
+}
+
+/// Helper function to ensure that the origin is either a regular signed account or root
+/// and returns the it as a `ContractOrigin`, if it is one of those.
+/// It errors otherwise.
+fn contract_ensure_signed_or_root<T: Config>(
+	origin: T::RuntimeOrigin,
+) -> Result<ContractOrigin<T>, BadOrigin> {
+	match ensure_signed_or_root(origin)? {
+		Some(t) => Ok(ContractOrigin::Signed(t)),
+		None => Ok(ContractOrigin::Root),
+	}
+}
+
+/// The type of callers supported by the contracts pallet.
+#[derive(Clone, Encode, Decode, PartialEq, TypeInfo)]
+pub enum Caller<T: Config> {
+	Account(T::AccountId),
+	Root,
+}
+
+impl<T: Config> Caller<T> {
+	/// Creates a new Signed ContractOrigin from an AccountId.
+	pub fn from_account_id(account_id: T::AccountId) -> Self {
+		Caller::Account(account_id)
+	}
+}
```

Comment:
same as above why not using Caller::Account directly

Reply:
same as above